### PR TITLE
fixed crash at mediaplayer closing, fixed tsmf event parameters-no video

### PR DESCRIPTION
**Pull request by AMV007**
freerdp crashed at media player closing windows 2008R2
no tsmf video output at test video sample about bunny.
all this fixed.

### CODE REVIEW

File: channels/tsmf/client/tsmf_media.c [hunk 1]
```diff
@@ -840,11 +854,13 @@ void tsmf_presentation_set_geometry_info(TSMF_PRESENTATION* presentation,
 	presentation->height = height;
 
 	tmp_rects = realloc(presentation->rects, sizeof(RDP_RECT) * num_rects);
-	if (!tmp_rects)
-		return;
```
Comment (bmiklautz) [inline channels/tsmf/client/tsmf_media.c]: Any specific reason why you removed this check?

File: channels/tsmf/client/tsmf_media.c [hunk 2]
```diff
@@ -375,6 +375,18 @@ static void tsmf_sample_playback_video(TSMF_SAMPLE* sample)
...
 
+		event.x = presentation->x;
+		event.y = presentation->y;
+		event.width = presentation->width;
+		event.height = presentation->height;
+
+		if (presentation->nr_rects > 0)
+		{
+			event.numVisibleRects = presentation->nr_rects;
+			event.visibleRects = (RECTANGLE_16*) calloc(1, event.numVisibleRects * sizeof(RECTANGLE_16));
```
Comment (bmiklautz) [inline channels/tsmf/client/tsmf_media.c]: result of  calloc isn't checked.

**Comment by bmiklautz** [comment]
@AMV007 how did you test (what tsfm "backend") and how did the crash happen?

**Comment by AMV007** [comment]
@bmiklautz 
>>Any specific reason why you removed this check?
it not removed, just moved down, because if realloc with size 0 we must update pointer presentation->rects to new value, because memory is freed !

Added check for calloc

**Comment by bmiklautz** [comment]
If a realloc with size 0 can happen this case should be handled separately otherwise you can't distinguish if realloc failed or if it was intended.

**Comment by bmiklautz** [comment]
@AMV007 I would also be interested in how you did your testing.

**Comment by AMV007** [comment]
fixed realloc and calloc.
Tested - simple win2008r2 - downloaded big_buck_bunny_720p_h264 video from description of tsmf and tried play it with media player.

**Comment by AMV007** [comment]
additionally fixed tsmf - no audio

File: channels/tsmf/client/tsmf_media.c [hunk 3]
```diff
@@ -406,6 +426,8 @@ static void tsmf_sample_playback_video(TSMF_SAMPLE* sample)
 			tsmf->FrameEvent(tsmf, &event);
 
 		free(event.frameData);
+
+		if(event.visibleRects!=NULL) free(event.visibleRects);
```
Comment (hardening) [inline channels/tsmf/client/tsmf_media.c]: coding style: spaces  around NULL, and free() on its own line

File: channels/tsmf/client/tsmf_media.c [hunk 4]
```diff
@@ -840,8 +875,10 @@ void tsmf_presentation_set_geometry_info(TSMF_PRESENTATION* presentation,
 	presentation->height = height;
 
 	tmp_rects = realloc(presentation->rects, sizeof(RDP_RECT) * num_rects);
-	if (!tmp_rects)
+
+	if (!tmp_rects&&num_rects)
```
Comment (hardening) [inline channels/tsmf/client/tsmf_media.c]: if num_rects is 0, presentation->rects is freed so it should be set to null

File: channels/tsmf/client/tsmf_media.c [hunk 5]
```diff
@@ -375,6 +377,24 @@ static void tsmf_sample_playback_video(TSMF_SAMPLE* sample)
...
+		{
+			event.numVisibleRects = presentation->nr_rects;
+			event.visibleRects = (RECTANGLE_16*) calloc(1, event.numVisibleRects * sizeof(RECTANGLE_16));
+			if (!event.visibleRects)
+			{
+				WLog_ERR(TAG, "can't allocate memory for copy rectangles");
+				return;
+			}
+			memcpy(event.visibleRects, presentation->rects, presentation->nr_rects * sizeof(RDP_RECT));
+			presentation->nr_rects=0;
```
Comment (hardening) [inline channels/tsmf/client/tsmf_media.c]: coding style: missing spaces around =

**Comment by hardening** [comment]
except my (blocking) remarks, +1

**Comment by AMV007** [comment]
fixed issues from hardening

**Comment by akallabeth** [comment]
@bmiklautz @hardening Any objections against a manual merge?

**Comment by hardening** [comment]
@akallabeth yep no pb

**Comment by akallabeth** [comment]
Superseeded by #3707

**Pull request closed without merging**